From 633b3837ab77445a4edb18812e2836d2ade5d6a5 Mon Sep 17 00:00:00 2001
From: Terry Zhou <bjzhou@marvell.com>
Date: Tue, 3 May 2016 16:46:10 +0800
Subject: [PATCH 0214/1345] ARM: mvebu: add core support for Msys SoC family

commit  66fcf58919b92adf259335e6ddd98be1dd90ee58 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

This commit adds the core support for Msys SoC's family:

- a new Kconfig option which selects the appropriate clock and
  pinctrl drivers as well as other common features (GIC, L2 cache,
  SMP, etc.). Most of them are shared with Armada 370/XP SoC's.
- a new DT_MACHINE_START which references the top-level compatible
  strings supported for the Marvell Msys machines.

Change-Id: Iaf169f8b0e9f08a6f3674eef835b48dd78d54efd
Signed-off-by: Terry Zhou <bjzhou@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/29695
Reviewed-by: Wilson Ding <dingwei@marvell.com>
Tested-by: Wilson Ding <dingwei@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../devicetree/bindings/arm/marvell/msys.txt       |   28 +++++++++++++++++
 arch/arm/mach-mvebu/Kconfig                        |   10 ++++++
 arch/arm/mach-mvebu/board-v7.c                     |   32 ++++++++++++++++++++
 arch/arm/mach-mvebu/cpu-reset.c                    |    4 +++
 4 files changed, 74 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/arm/marvell/msys.txt

diff --git a/Documentation/devicetree/bindings/arm/marvell/msys.txt b/Documentation/devicetree/bindings/arm/marvell/msys.txt
new file mode 100644
index 0000000..ad3a659
--- /dev/null
+++ b/Documentation/devicetree/bindings/arm/marvell/msys.txt
@@ -0,0 +1,28 @@
+Marvell Msys Platforms Device Tree Bindings
+---------------------------------------------------------------
+
+Boards with a SoC of the Marvell Msys families
+shall have the following property:
+
+Required root node property:
+
+compatible: must contain "marvell,armada-370-xp"
+
+In addition, boards using the Marvell Msys SoC shall have the
+following property:
+
+Required root node property:
+
+compatible: must contain "marvell,msys"
+
+In addition, boards using the Marvell Msys AlleyCat3 SoC shall have the
+following property:
+
+Required root node property:
+
+compatible: must contain "marvell,msys-ac3"
+
+Example:
+
+compatible = "marvell,msys-ac3-db", "marvell,msys-ac3",
+		     "marvell,msys", "marvell,armada-370-xp";
diff --git a/arch/arm/mach-mvebu/Kconfig b/arch/arm/mach-mvebu/Kconfig
index 541647f..8b8309e 100644
--- a/arch/arm/mach-mvebu/Kconfig
+++ b/arch/arm/mach-mvebu/Kconfig
@@ -130,4 +130,14 @@ config MACH_KIRKWOOD
 	  Say 'Y' here if you want your kernel to support boards based
 	  on the Marvell Kirkwood device tree.
 
+config MACH_MSYS
+	bool "Marvell Msys boards"
+	select MSYS_CLK
+	select CPU_PJ4B
+	select MACH_MVEBU_V7
+	select PINCTRL_MSYS
+	help
+	  Say 'Y' here if you want your kernel to support boards based
+	  on the Marvell Msys SoC family with device tree.
+
 endif
diff --git a/arch/arm/mach-mvebu/board-v7.c b/arch/arm/mach-mvebu/board-v7.c
index da144d0..4650c3f 100644
--- a/arch/arm/mach-mvebu/board-v7.c
+++ b/arch/arm/mach-mvebu/board-v7.c
@@ -156,6 +156,16 @@ static void __init mvebu_init_irq(void)
 	BUG_ON(mvebu_mbus_dt_init(coherency_available()));
 }
 
+static void __init msys_irqchip_init(void)
+{
+	/* Because the switch interrupt driver (marvell,swic) uses register from
+	* the switch region space, the decoding window for switch must be
+	* initialized, before calling interrupt drivers.
+	*/
+	BUG_ON(mvebu_mbus_dt_init(coherency_available()));
+	mvebu_init_irq();
+}
+
 static void __init i2c_quirk(void)
 {
 	struct device_node *np;
@@ -253,3 +263,25 @@ static void __init mvebu_dt_init(void)
 	.restart	= mvebu_restart,
 	.dt_compat	= armada_39x_dt_compat,
 MACHINE_END
+
+static const char * const msys_dt_compat[] __initconst = {
+	"marvell,msys",
+	NULL,
+};
+
+
+DT_MACHINE_START(MSYS_DT, "Marvell SYS (Device Tree)")
+	.l2c_aux_val	= 0,
+	.l2c_aux_mask	= ~0,
+/*
+ * The following field (.smp) is still needed to ensure backward
+ * compatibility with old Device Trees that were not specifying the
+ * cpus enable-method property.
+ */
+	.smp		= smp_ops(armada_xp_smp_ops),
+	.init_machine	= mvebu_dt_init,
+	.init_irq       = msys_irqchip_init,
+	.restart	= mvebu_restart,
+	.reserve        = mvebu_memblock_reserve,
+	.dt_compat	= msys_dt_compat,
+MACHINE_END
diff --git a/arch/arm/mach-mvebu/cpu-reset.c b/arch/arm/mach-mvebu/cpu-reset.c
index f33a31c..1928910 100644
--- a/arch/arm/mach-mvebu/cpu-reset.c
+++ b/arch/arm/mach-mvebu/cpu-reset.c
@@ -85,6 +85,10 @@ static int __init mvebu_cpu_reset_init(void)
 		 */
 		np = of_find_compatible_node(NULL, NULL,
 					     "marvell,armada-370-xp-pmsu");
+		if (!np)
+			np = of_find_compatible_node(NULL, NULL,
+						"marvell,msys-pmsu");
+
 		if (np) {
 			pr_warn(FW_WARN "deprecated pmsu binding\n");
 			res_idx = 1;
-- 
1.7.9.5

